File: mailnews/base/src/UrlListener.cpp
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "UrlListener.h"

NS_IMPL_ISUPPORTS(UrlListener, nsIUrlListener)

NS_IMETHODIMP UrlListener::OnStartRunningUrl(nsIURI* url) {
  if (!mStartFn) {
    return NS_OK;
  }
  return mStartFn(url);
}

NS_IMETHODIMP UrlListener::OnStopRunningUrl(nsIURI* url, nsresult exitCode) {
  if (!mStopFn) {
    return NS_OK;
  }
  return mStopFn(url, exitCode);
}

NS_IMPL_ISUPPORTS(CopyServiceListener, nsIMsgCopyServiceListener)

NS_IMETHODIMP CopyServiceListener::OnStartCopy() {
  if (!mStartFn) {
    return NS_OK;
  }
  return mStartFn();
}

NS_IMETHODIMP CopyServiceListener::OnProgress(uint32_t progress,
                                              uint32_t progressMax) {
  return NS_OK;
}

NS_IMETHODIMP CopyServiceListener::SetMessageKey(nsMsgKey key) { return NS_OK; }

NS_IMETHODIMP CopyServiceListener::GetMessageId(nsACString& messageId) {
  return NS_ERROR_NOT_IMPLEMENTED;
}

NS_IMETHODIMP CopyServiceListener::OnStopCopy(nsresult status) {
  if (!mStopFn) {
    return NS_OK;
  }
  return mStopFn(status);
}

File: rust/ews_xpcom/src/safe_xpcom/url_listener.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

use nserror::nsresult;
use xpcom::{
    interfaces::{nsIURI, nsIUrlListener},
    RefPtr,
};

use crate::client::XpComEwsError;

use super::{SafeListener, SafeListenerWrapper, SafeUri};

/// See [`SafeListenerWrapper`].
pub(crate) type SafeUrlListener = SafeListenerWrapper<nsIUrlListener>;

impl SafeUrlListener {
    /// Convert types and forward to [`nsIUrlListener::OnStartRunningUrl`].
    pub fn on_start_running_url(&self, uri: SafeUri) -> nsresult {
        let uri: RefPtr<nsIURI> = uri.into();
        // SAFETY: uri points to a valid nsIRUI
        unsafe { self.0.OnStartRunningUrl(&*uri) }
    }

    /// Convert types and forward to [`nsIUrlListener::OnStopRunningUrl`]. This
    /// is invoked by [`Self::on_success`] and [`Self::on_failure`].
    fn on_stop_running_url(&self, uri: SafeUri, status: nsresult) -> nsresult {
        let uri: RefPtr<nsIURI> = uri.into();
        // SAFETY: uri points to a valid nsIRUI
        unsafe { self.0.OnStopRunningUrl(&*uri, status) }
    }
}

impl SafeListener for SafeUrlListener {
    type OnSuccessArg = SafeUri;
    type OnFailureArg = SafeUri;

    /// Calls [`nsIUrlListener::OnStopRunningUrl`] with the appropriate
    /// arguments.
    fn on_success(&self, uri: SafeUri) -> Result<(), nsresult> {
        self.on_stop_running_url(uri, nserror::NS_OK).to_result()?;
        Ok(())
    }

    /// Calls [`nsIUrlListener::OnStopRunningUrl`] with the appropriate
    /// arguments.
    fn on_failure(&self, err: &XpComEwsError, uri: SafeUri) -> Result<(), nsresult> {
        self.on_stop_running_url(uri, err.into()).to_result()?;
        Ok(())
    }
}
